### PR TITLE
storage: extend the `ExternalStorage` interface and implements

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
extend the `ExternalStorage` interface and implement the new methods for local and s3 storage. This extension is currently used  in lightning to support restore from s3 storage.

### What is changed and how it works?
Add new methods for ExternalStorage:
```
type ReadSeekCloser interface {
	io.Reader
	io.Seeker
	io.Closer
}

type ExtrenalStorage interface {
	// Open a Reader by file name
	Open(ctx context.Context, name string) (ReadSeekCloser, error)
	// WalkDir traverse all the files in a dir
	WalkDir(ctx context.Context, fn func(string, int64) error) error
}
```

Since s3 GetObject does not provide seek directly, we implement this by: 1) read and skip the bytes if the distance is small, else we open a new reader set the start position to the seek position. 

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

Related changes

### Release Note

 - No release note

### CODE REVIEW

File: pkg/storage/s3.go [hunk 1]
```diff
@@ -292,3 +299,143 @@ func (rs *S3Storage) FileExists(ctx context.Context, file string) (bool, error)
...
+}
+
+// Close implement the io.Closer interface
+func (r *s3ObjectReader) Close() error {
+	return r.reader.Close()
+}
+
+// Seek implement the io.Seeker interface
+func (r *s3ObjectReader) Seek(offset int64, whence int) (int64, error) {
+	// if seek ahead no more than 64k, we read add drop these data
```
Comment (kennytm) [inline pkg/storage/s3.go:398]: move this comment to line 413.

File: pkg/storage/s3.go [hunk 2]
```diff
@@ -292,3 +299,143 @@ func (rs *S3Storage) FileExists(ctx context.Context, file string) (bool, error)
...
+	var realOffset int64
+	if whence == io.SeekStart {
+		realOffset = offset
+	} else if whence == io.SeekCurrent {
+		realOffset = r.pos + offset
+	} else {
+		// TODO
+		return 0, errors.New("seek by SeekEnd is not supported yet")
+	}
```
Comment (kennytm) [inline pkg/storage/s3.go:407]:
```suggestion
	switch whence {
	case io.SeekStart:
		realOffset = offset
	case io.SeekCurrent:
		realOffset = r.pos + offset
	default:
		// TODO
		return 0, errors.New("seek by SeekEnd is not supported yet")
	}
```

File: pkg/storage/s3.go [hunk 3]
```diff
@@ -292,3 +299,143 @@ func (rs *S3Storage) FileExists(ctx context.Context, file string) (bool, error)
...
+		buf := make([]byte, batch, batch)
+		total := offset - r.pos
+		remain := total
+		for remain > 0 {
+			n, err := r.Read(buf[:utils.MinInt64(remain, batch)])
+			if err != nil {
+				return total - remain + int64(n), err
+			}
+			remain -= int64(n)
+		}
```
Comment (kennytm) [inline pkg/storage/s3.go:424]:
```suggestion
		written, err := io.CopyN(ioutil.Discard, r, offset-r.pos)
		if err != nil {
			return r.pos + written, err
		}
```

File: pkg/storage/gcs.go [hunk 4]
```diff
@@ -123,6 +123,18 @@ func (s *gcsStorage) FileExists(ctx context.Context, name string) (bool, error)
...
 }
 
+// Open a Reader by file name
+func (s *gcsStorage) Open(ctx context.Context, name string) (ReadSeekCloser, error) {
+	// TODO, implement this if needed
+	panic("Unsupported Operation")
+}
+
+// WalkDir traverse all the files in a dir
+func (s *gcsStorage) WalkDir(ctx context.Context, fn func(string, int64) error) error {
```
Comment (kennytm) [inline pkg/storage/gcs.go:133]: please document that the first argument of `fn` is the full path to be passed into `Open()` and the second argument is the file size.

File: pkg/storage/flags.go [hunk 5]
```diff
@@ -8,13 +8,13 @@ import (
 
 // DefineFlags adds flags to the flag set corresponding to all backend options.
 func DefineFlags(flags *pflag.FlagSet) {
-	defineS3Flags(flags)
+	DefineS3Flags(flags)
 	defineGCSFlags(flags)
 }
```
Comment (kennytm) [inline pkg/storage/flags.go:13]: what is the rationale of only exposing S3 but not GCS?

**Comment by glorv** [comment]
@kennytm  Shall we change the `s3` and `gcs` flag definitions from `pflag` to `flag` so that lightning can share the same definitions without migrating into `pflag`?

**Comment by kennytm** [comment]
@glorv we don't really need those flags since they can be passed via URL parameters (`s3://bucket/prefix?region=us-east-1` etc).

File: pkg/storage/s3.go [hunk 6]
```diff
@@ -292,3 +300,143 @@ func (rs *S3Storage) FileExists(ctx context.Context, file string) (bool, error)
...
+	if err != nil {
+		return nil, err
+	}
+
+	// FIXME: we test in minio, when request with Range, the result.AcceptRanges is a bare string 'range',
+	//  not sure whether this is a feature or bug
+	//if rangeOffset != nil && (result.AcceptRanges == nil || *result.AcceptRanges != *rangeOffset) {
+	//	return nil, errors.Errorf("open file '%s' failed, expected range: %s, got: %v",
+	//		name, *rangeOffset, result.AcceptRanges)
+	//}
```
Comment (overvenus) [inline pkg/storage/s3.go:378]: Have we tested in AWS S3?

File: pkg/storage/local.go [hunk 7]
```diff
@@ -42,7 +72,10 @@ func pathExists(_path string) (bool, error) {
 	return true, nil
 }
 
-func newLocalStorage(base string) (*localStorage, error) {
+// NewLocalStorage return a LocalStorage at directory `base`.
+//
+// export for test.
+func NewLocalStorage(base string) (*LocalStorage, error) {
```
Comment (overvenus) [inline pkg/storage/local.go:78]: Where is the test?

File: pkg/storage/s3.go [hunk 8]
```diff
@@ -292,3 +300,143 @@ func (rs *S3Storage) FileExists(ctx context.Context, file string) (bool, error)
...
+	default:
+		// TODO: maybe we can fetch the object stat and calculate the absolute offset
+		return 0, errors.New("seek by SeekEnd is not supported yet")
+	}
+
+	if realOffset == r.pos {
+		return realOffset, nil
+	}
+
+	// if seek ahead no more than 64k, we read add drop these data
```
Comment (overvenus) [inline pkg/storage/s3.go:420]:
```suggestion
	// if seek ahead no more than 64k, we discard these data
```

File: pkg/storage/s3.go [hunk 9]
```diff
@@ -292,3 +300,143 @@ func (rs *S3Storage) FileExists(ctx context.Context, file string) (bool, error)
...
+		return realOffset, nil
+	}
+
+	// close current read and open a new one which target offset
+	err := r.reader.Close()
+	if err != nil {
+		return 0, err
+	}
+
+	newReader, err := r.storage.open(context.Background(), r.name, realOffset, 0)
```
Comment (overvenus) [inline pkg/storage/s3.go:435]: Please avoid use Background context.

File: pkg/storage/local.go (hunk 7 above)
Comment (glorv) [reply pkg/storage/local.go:78]: used in lightning tests

**Comment by overvenus** [comment]
LGTM

**Pull request merged**